src/simple_http.rs

```diff
@@ -73,29 +75,50 @@ impl SimpleHttpTransport {
     }
 
     fn request<R>(&self, req: impl serde::Serialize) -> Result<R, Error>
+    where
+        R: for<'a> serde::de::Deserialize<'a>,
+    {
+        match self.try_request(req) {
+            Ok(response) => Ok(response),
+            Err(err) => {
+                *self.sock.lock().unwrap() = None;
+                Err(err)
+            }
+        }
+    }
+
+    fn try_request<R>(&self, req: impl serde::Serialize) -> Result<R, Error>
     where
         R: for<'a> serde::de::Deserialize<'a>,
     {
         // Open connection
         let request_deadline = Instant::now() + self.timeout;
-        #[cfg(feature = "proxy")]
-        let mut sock = if let Some((username, password)) = &self.proxy_auth {
-            Socks5Stream::connect_with_password(
-                self.proxy_addr,
-                self.addr,
-                username.as_str(),
-                password.as_str(),
-            )?
-            .into_inner()
-        } else {
-            Socks5Stream::connect(self.proxy_addr, self.addr)?.into_inner()
-        };
-
-        #[cfg(not(feature = "proxy"))]
-        let mut sock = TcpStream::connect_timeout(&self.addr, self.timeout)?;
+        let mut sock = self.sock.lock().unwrap();
+        if sock.is_none() {
+            #[cfg(feature = "proxy")]
+            {
+                *sock = Some(if let Some((username, password)) = &self.proxy_auth {
+                    Socks5Stream::connect_with_password(
+                        self.proxy_addr,
+                        self.addr,
+                        username.as_str(),
+                        password.as_str(),
+                    )?
+                    .into_inner()
+                } else {
+                    Socks5Stream::connect(self.proxy_addr, self.addr)?.into_inner()
+                });
+            }
 
-        sock.set_read_timeout(Some(self.timeout))?;
-        sock.set_write_timeout(Some(self.timeout))?;
+            #[cfg(not(feature = "proxy"))]
+            {
+                let stream = TcpStream::connect_timeout(&self.addr, Duration::from_secs(15))?;
+                stream.set_read_timeout(Some(self.timeout))?;
+                stream.set_write_timeout(Some(self.timeout))?;
+                *sock = Some(stream);
+            }
+        }
```

Comment:
Could you refactor this code to instead look like
```
let stream =
#[cfg(feature = "proxy")] {
    // ...
};
#[cfg(not(feature = "proxy"))] {
    // ...
};
*sock = Some(stream);
```
So that both cases have the same side-effect profile and it's clear that in both cases you're constructing a new object and assigning it to `*sock`.

Reply:
I believe using cfg with expressions is experimental so I can't do it that nicely yet. I have made it look a bit cleaner in the updated commit.